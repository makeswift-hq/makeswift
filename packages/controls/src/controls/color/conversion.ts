import parseColor from 'color'

import { type Swatch } from '../../resources'

export function swatchToColorString(
  swatch: Swatch | null,
  alpha: number,
  defaultValue?: string,
): string | undefined {
  if (swatch == null) {
    return defaultValue === undefined
      ? undefined
      : safeParseColor(defaultValue).rgb().string()
  }

  return parseColor({
    h: swatch.hue,
    s: swatch.saturation,
    l: swatch.lightness,
  })
    .alpha(alpha)
    .rgb()
    .string()
}

function safeParseColor(value: string) {
  try {
    return parseColor(value)
  } catch {
    return parseColor()
  }
}
