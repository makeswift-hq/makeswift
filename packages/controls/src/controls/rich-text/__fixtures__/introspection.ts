import { RichTextDTO } from '../dto'

export const SWATCH_ID = 'SWATCH_ID='
export const TYPOGRAPHY_ID = 'TYPOGRAPHY_ID='
export const PAGE_ID = 'PAGE_ID='

export const introspection: RichTextDTO = {
  object: 'value',
  annotations: undefined,
  data: undefined,
  document: {
    data: undefined,
    nodes: [
      {
        data: {},
        type: 'blockquote',
        nodes: [
          {
            object: 'inline',
            type: 'link',
            data: {
              type: 'OPEN_PAGE',
              payload: {
                pageId: PAGE_ID,
                openInNewTab: false,
              },
            },
            nodes: [
              {
                text: 'Credibly implement global synergy, then collaboratively implement goal-oriented sprints. Rapidiously initiate standards compliant clouds. ',
                marks: [
                  {
                    data: {
                      value: {
                        id: TYPOGRAPHY_ID,
                        style: [
                          {
                            value: {
                              color: {
                                swatchId: SWATCH_ID,
                                alpha: 1,
                              },
                            },
                          },
                          {
                            value: {
                              fontWeight: 500,
                            },
                          },
                          {
                            value: {
                              color: {
                                swatchId: undefined,
                                alpha: 1,
                              },
                            },
                          },
                        ],
                      },
                    },
                    type: 'typography',
                    object: 'mark',
                  },
                ],
                object: 'text',
              },
            ],
          },
        ],
        object: 'block',
      },
    ],
    object: 'document',
  },
  selection: {
    focus: {
      path: [1, 2],
      object: 'point',
      offset: 0,
    },
    marks: undefined,
    anchor: {
      path: [0, 0],
      object: 'point',
      offset: 0,
    },
    object: 'selection',
    isFocused: false,
  },
}
