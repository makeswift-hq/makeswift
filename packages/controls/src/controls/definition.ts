import { z } from 'zod'

import { type ParseResult } from '../lib/zod'

import { type Data } from '../common/types'
import {
  type CopyContext,
  type MergeContext,
  type MergeTranslatableDataContext,
} from '../context'
import { type IntrospectionTarget } from '../introspection'
import {
  Serializable,
  serializeObject,
  type SerializedRecord,
} from '../serialization'

import { ControlInstance, type SendMessage } from './instance'

export type SchemaType<T> = z.ZodType<T>
export type SchemaTypeAny = SchemaType<any> | z.ZodBranded<SchemaType<any>, any>

export type Resolvable<T> = {
  readStableValue(previous?: T): T
  subscribe(onUpdate: () => void): () => void
}

// workaround for TS type inference issues
export type UndefinedOr<T> = T | undefined

export abstract class ControlDefinition<
  ControlType extends string = string,
  Config = unknown,
  DataType extends Data = Data,
  ValueType extends Data = Data,
  ResolvedValueType = Data | unknown,
  InstanceType extends ControlInstance<any> = ControlInstance<any>,
> extends Serializable {
  constructor(readonly config: Config) {
    super()
  }

  abstract get controlType(): ControlType

  abstract get schema(): {
    definition: SchemaType<unknown>
    type: SchemaType<ControlType>
    data: SchemaType<DataType>
    value: SchemaType<ValueType>
    resolvedValue: SchemaType<ResolvedValueType>
  }

  abstract safeParse(
    data: unknown | undefined,
  ): ParseResult<UndefinedOr<DataType>>

  abstract fromData(data: UndefinedOr<DataType>): UndefinedOr<ValueType>
  abstract toData(value: ValueType): DataType

  abstract copyData(
    data: UndefinedOr<DataType>,
    context: CopyContext,
  ): UndefinedOr<DataType>

  mergeData(
    base: DataType,
    override: UndefinedOr<DataType>,
    _context: MergeContext,
  ): DataType {
    return override ?? base
  }

  getTranslatableData(_data: DataType): Data {
    return null
  }

  mergeTranslatedData(
    data: DataType,
    _translatedData: Data,
    _context: MergeTranslatableDataContext,
  ): Data {
    return data as Data
  }

  resolveValue(
    _data: UndefinedOr<DataType>,
  ): Resolvable<UndefinedOr<ResolvedValueType>> {
    console.assert(
      false,
      `${this.controlType}: 'resolveValue' is not implemented`,
    )

    return {
      readStableValue: (_previous?: ResolvedValueType) => undefined,
      subscribe: () => () => {},
    }
  }

  abstract createInstance(sendMessage: SendMessage<any>): InstanceType

  abstract serialize(): [SerializedRecord, Transferable[]]

  introspect<R>(
    data: UndefinedOr<DataType>,
    target: IntrospectionTarget<R>,
  ): R[] {
    return target.introspect(data)
  }
}

export function serialize(
  config: unknown,
  rest: { type: string } & Record<string, unknown>,
): [SerializedRecord, Transferable[]] {
  const [serializedConfig, transferables] = serializeObject(config)
  return [
    {
      config: serializedConfig,
      ...rest,
    } as unknown as SerializedRecord,
    transferables,
  ]
}
