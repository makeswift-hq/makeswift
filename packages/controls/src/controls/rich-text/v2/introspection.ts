import { type IntrospectionTarget } from '../../../introspection'

import * as Slate from '../slate'

import { RichTextPluginControl } from './plugin'

export function introspectNodes<R>(
  nodes: Slate.Descendant[],
  target: IntrospectionTarget<R>,
  pluginControls: RichTextPluginControl[],
): R[] {
  return nodes.flatMap(introspectNode)

  function introspectNode(descendant: Slate.Descendant): R[] {
    if (Slate.Element.isElement(descendant)) {
      return [
        ...introspectElement(descendant),
        ...descendant.children.flatMap(introspectNode),
      ]
    }

    if (Slate.Text.isText(descendant)) {
      return introspectText(descendant)
    }

    return []
  }

  function introspectElement(descendant: Slate.Element) {
    return pluginControls.flatMap((control) =>
      control.getElementValue
        ? control.definition.introspect(
            control.getElementValue(descendant),
            target,
          )
        : [],
    )
  }

  function introspectText(text: Slate.Text) {
    return pluginControls.flatMap((control) =>
      control.getLeafValue
        ? control.definition.introspect(control.getLeafValue(text), target)
        : [],
    )
  }
}
