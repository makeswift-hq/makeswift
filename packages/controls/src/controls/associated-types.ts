import { ControlDefinition } from './definition'

export type ControlType<D> =
  D extends ControlDefinition<infer ControlType, any, any, any, any, any>
    ? ControlType
    : never

export type ConfigType<D> =
  D extends ControlDefinition<any, infer Config, any, any, any, any>
    ? Config
    : never

export type DataType<D> =
  D extends ControlDefinition<any, any, infer DataType, any, any, any>
    ? DataType
    : never

export type ValueType<D> =
  D extends ControlDefinition<any, any, any, infer ValueType, any, any>
    ? ValueType
    : never

export type ResolvedValueType<D> =
  D extends ControlDefinition<any, any, any, any, infer ResolvedValueType, any>
    ? ResolvedValueType
    : never

export type InstanceType<D> =
  D extends ControlDefinition<any, any, any, any, any, infer InstanceType>
    ? InstanceType
    : never
