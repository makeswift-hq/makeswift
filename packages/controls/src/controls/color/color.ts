import { match } from 'ts-pattern'
import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey, Schema } from '../../common'
import { type CopyContext } from '../../context'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import { ControlDefinition, serialize, type SchemaType } from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

type Config = z.infer<typeof Definition.schema.relaxed.config>

type SchemaByDefaultValue<D extends Config['defaultValue']> =
  undefined extends D
    ? typeof Definition.schema.relaxed
    : typeof Definition.schema.strict

type Schema<C extends Config> = SchemaByDefaultValue<C['defaultValue']>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>

type ReturnedSchemaType<C extends Config> = {
  definition: typeof Definition.schema.relaxed.definition
  type: typeof Definition.schema.relaxed.type
  data: SchemaType<DataType<C>>
  value: SchemaType<ValueType<C>>
  resolvedValue: SchemaType<ResolvedValueType<C>>
}

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'color::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::color' as const

  static get schema() {
    const type = z.literal(this.type)
    const version = z.literal(1).optional()
    const value = Schema.colorData
    const resolvedSwatchValue = Schema.resolvedColorData

    const schemas = <R>(resolvedValue: SchemaType<R>) => {
      const data = value.merge(
        z.object({
          [ControlDataTypeKey]: z.literal(this.v1DataType).optional(),
        }),
      )

      const config = z.object({
        label: z.string().optional(),
        labelOrientation: z
          .union([z.literal('horizontal'), z.literal('vertical')])
          .optional(),
        defaultValue: resolvedValue,
        hideAlpha: z.boolean().optional(),
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return {
        type,
        value,
        resolvedSwatchValue,
        resolvedValue,
        data,
        config,
        version,
        definition,
      }
    }

    return {
      version,
      relaxed: schemas(z.string().optional()),
      strict: schemas(z.string()),
    }
  }

  static deserialize(data: DeserializedRecord): ColorDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Color deserialization: expected '${Definition.type}', got '${data.type}'`,
      )

    const { config, version } = Definition.schema.relaxed.definition.parse(data)
    return new ColorDefinition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema(): ReturnedSchemaType<C> {
    return Definition.schema.relaxed
  }

  get dataSchema() {
    return (
      (this.config.defaultValue === undefined
        ? Definition.schema.relaxed
        : Definition.schema.strict) as Schema<C>
    ).data
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ swatchId, alpha }) => ({
        swatchId,
        alpha,
      }))
      .otherwise((value) => value)
  }

  toData(value: ValueType<C>): DataType<C> {
    return match(this.version)
      .with(1, () => ({
        ...Definition.dataSignature.v1,
        ...value,
      }))
      .with(undefined, () => value)
      .exhaustive()
  }

  copyData(
    data: DataType<C> | undefined,
    { replacementContext }: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return data

    const replaceSwatchId = (swatchId: string) =>
      replacementContext.swatchIds.get(swatchId) ?? swatchId

    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, (val) => ({
        ...val,
        swatchId: replaceSwatchId(val.swatchId),
      }))
      .otherwise((val) => ({
        ...val,
        swatchId: replaceSwatchId(val.swatchId),
      }))
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
      version: this.version,
    })
  }
}

export class ColorDefinition<C extends Config = Config> extends Definition<C> {}

type UserConfig<D extends Config['defaultValue']> = Config & {
  defaultValue?: D
}

type NormedConfig<D extends Config['defaultValue']> = z.infer<
  SchemaByDefaultValue<D>['config']
>

export function Color<D extends Config['defaultValue']>(
  config?: UserConfig<D>,
): ColorDefinition<NormedConfig<D>> {
  return new ColorDefinition<NormedConfig<D>>(config ?? {}, 1)
}
