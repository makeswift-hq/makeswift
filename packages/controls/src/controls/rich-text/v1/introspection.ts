import { type IntrospectionTarget } from '../../../introspection'
import { type DataType } from '../../associated-types'
import { GenericLink, LinkDefinition } from '../../link'
import { unstable_Typography } from '../../typography'

import { type RichTextDTO } from '../dto'
import {
  TextJSON,
  type InlineJSON,
  type MarkJSON,
  type NodeJSON,
} from '../dto/types'

export function introspectRichTextData<R>(
  data: RichTextDTO | undefined,
  target: IntrospectionTarget<R>,
): R[] {
  if (data == null || data.document == null) return []

  function introspectNode(node: NodeJSON): R[] {
    switch (node.object) {
      case 'document':
      case 'block':
        return node.nodes?.flatMap(introspectNode) ?? []

      case 'inline':
      case 'text':
        return introspectInlineOrTextNode(node)

      default:
        return []
    }
  }

  function introspectInlineOrTextNode(node: InlineJSON | TextJSON): R[] {
    switch (node.object) {
      case 'inline':
        return [
          ...introspectInlineNode(node),
          ...(node.nodes?.flatMap(introspectInlineOrTextNode) ?? []),
        ]

      case 'text':
        return node.marks?.flatMap(introspectMark) ?? []

      default:
        return []
    }
  }

  function introspectInlineNode(inline: InlineJSON): R[] {
    switch (inline.type) {
      case 'link': {
        return GenericLink().introspect(
          inline.data as DataType<LinkDefinition>,
          target,
        )
      }

      default:
        return []
    }
  }

  function introspectMark(mark: MarkJSON): R[] {
    const typographyResult = unstable_Typography().introspect(
      mark.data?.value,
      target,
    )

    return [...typographyResult]
  }

  return introspectNode(data.document)
}
