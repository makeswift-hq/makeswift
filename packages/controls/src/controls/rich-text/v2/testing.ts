import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
