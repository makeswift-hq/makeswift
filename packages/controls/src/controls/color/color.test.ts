import { testDefinition } from '../../testing/test-definition'

import { createReplacementContext, type CopyContext } from '../../context'
import { Targets } from '../../introspection'

import { type ValueType } from '../associated-types'
import { ControlDefinition } from '../definition'

import { Color, ColorDefinition } from './color'

function testColorDefinition<Def extends ControlDefinition>(
  definition: Def,
  values: ValueType<Def>[],
  invalidValues: unknown[],
) {
  testDefinition(definition, values, invalidValues)

  describe('copyData', () => {
    const context: CopyContext = {
      replacementContext: createReplacementContext({
        swatchIds: { '[swatch-id-1]': '[swatch-id-replaced]' },
      }),
      copyElement: (node) => node,
    }

    test('replaces `swatchId` found in replacement context', () => {
      // Act
      const result = definition.copyData(
        { swatchId: '[swatch-id-1]', alpha: 0.75 },
        context,
      )

      // Assert
      expect(result).toEqual({ swatchId: '[swatch-id-replaced]', alpha: 0.75 })
    })

    test('does not replace `swatchId` not found in replacement context', () => {
      // Act
      const result = definition.copyData(
        { swatchId: '[swatch-id-2]', alpha: 1 },
        context,
      )

      // Assert
      expect(result).toEqual({ swatchId: '[swatch-id-2]', alpha: 1 })
    })

    test.each([null, undefined])('gracefully handles %s', (value) => {
      expect(definition.copyData(value, context)).toBe(value)
    })
  })

  describe('introspect swatch IDs', () => {
    test("returns data's `swatchId`", () => {
      expect(
        definition.introspect({ swatchId: '[swatch-id-2]' }, Targets.Swatch),
      ).toEqual(['[swatch-id-2]'])
    })

    test.each([null, undefined])('gracefully handles %s', (value) => {
      expect(definition.introspect(value, Targets.Swatch)).toEqual([])
    })
  })
}

describe('Color', () => {
  describe('constructor', () => {
    test.each(['red', undefined])(
      'call with default value `%s` returns versioned definition',
      (value) => {
        expect(
          Color({
            label: 'color',
            defaultValue: value,
          }),
        ).toMatchSnapshot()
      },
    )

    test('disallows extraneous properties', () => {
      Color({
        label: 'background',
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: ColorDefinition) {}
    assignTest(Color({ label: 'background' }))
    assignTest(Color({ defaultValue: 'red' }))
    assignTest(Color({ label: 'background', labelOrientation: 'horizontal' }))
    assignTest(
      Color({
        label: 'background',
        labelOrientation: undefined,
        hideAlpha: true,
        defaultValue: '#fff' as string,
      }),
    )

    assignTest(
      Color({
        label: undefined,
        labelOrientation: undefined,
        hideAlpha: undefined,
        defaultValue: undefined,
      }),
    )
  })

  const invalidValues = [null, 17, 'text', { swatchId: 42 }, { alpha: 1 }]

  testColorDefinition(
    Color({ defaultValue: 'red', label: 'visible' }),
    [
      { swatchId: '[swatch-id-1]', alpha: 1 },
      { swatchId: '[swatch-id-2]', alpha: 0.5 },
    ],
    invalidValues,
  )
})
