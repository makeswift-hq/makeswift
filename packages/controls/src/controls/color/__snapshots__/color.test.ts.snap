// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color constructor call with default value \`red\` returns versioned definition 1`] = `
ColorDefinition {
  "config": {
    "defaultValue": "red",
    "label": "color",
  },
  "version": 1,
}
`;

exports[`Color constructor call with default value \`undefined\` returns versioned definition 1`] = `
ColorDefinition {
  "config": {
    "defaultValue": undefined,
    "label": "color",
  },
  "version": 1,
}
`;

exports[`Color definition w/ config {"defaultValue":"red","label":"visible"} toData returns v1 data for \`{ swatchId: '[swatch-id-1]', alpha: 1 }\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "color::v1",
  "alpha": 1,
  "swatchId": "[swatch-id-1]",
}
`;

exports[`Color definition w/ config {"defaultValue":"red","label":"visible"} toData returns v1 data for \`{ swatchId: '[swatch-id-2]', alpha: 0.5 }\` when definition version is 1 1`] = `
{
  "@@makeswift/type": "color::v1",
  "alpha": 0.5,
  "swatchId": "[swatch-id-2]",
}
`;
